servlet/src/main/java/io/undertow/servlet/handlers/security/ServletFormAuthenticationMechanism.java

```diff
@@ -230,33 +230,57 @@ protected void storeInitialLocation(final HttpServerExchange exchange, byte[] by
 
     @Override
     protected void handleRedirectBack(final HttpServerExchange exchange) {
-        final ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-        HttpServletResponse resp = (HttpServletResponse) servletRequestContext.getServletResponse();
-        HttpSessionImpl httpSession = servletRequestContext.getCurrentServletContext().getSession(exchange, false);
-        if (httpSession != null) {
-            Session session;
-            if (System.getSecurityManager() == null) {
-                session = httpSession.getSession();
-            } else {
-                session = AccessController.doPrivileged(new HttpSessionImpl.UnwrapSessionAction(httpSession));
-            }
-            Integer originalSessionTimeout = (Integer) session.removeAttribute(ORIGINAL_SESSION_TIMEOUT);
-            if (originalSessionTimeout != null) {
-                session.setMaxInactiveInterval(originalSessionTimeout);
-            }
+        final Session session = getAndInitializeSession(exchange, false);
+        if (session != null) {
             String path = (String) session.getAttribute(SESSION_KEY);
             if ((path == null || overrideInitial) && defaultPage != null) {
                 path = defaultPage;
             }
             if (path != null) {
+                final ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+                final HttpServletResponse resp = (HttpServletResponse) servletRequestContext.getServletResponse();
                 try {
                     resp.sendRedirect(path);
                 } catch (IOException e) {
                     throw new RuntimeException(e);
                 }
             }
         }
+    }
+
+    private Session getAndInitializeSession(final HttpServerExchange exchange, final boolean createNewSession) {
+        final ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+        final ServletContextImpl servletContextImpl =  servletRequestContext.getCurrentServletContext();
+        HttpSessionImpl httpSession = servletContextImpl.getSession(exchange, false);
+        if (httpSession == null && !createNewSession) return null;
+
+        boolean newSession = false;
+        if (httpSession == null) {
+            httpSession = servletContextImpl.getSession(exchange, true);
+            newSession = true;
+        }
+
+        Session session;
+        if (System.getSecurityManager() == null) {
+            session = httpSession.getSession();
+        } else {
+            session = AccessController.doPrivileged(new HttpSessionImpl.UnwrapSessionAction(httpSession));
+        }
+
+        if (newSession) {
+            final int originalMaxInactiveInterval = session.getMaxInactiveInterval();
+            if (originalMaxInactiveInterval > authenticationSessionTimeout) {
```

Comment:
just a minor pedantic adjustment? super.authenticationSessionTimeout ?

Reply:
Though Im not hell bent on that, its just good practice IMHO.

Reply:
okay, agreed! Because this is minor, though, I am merging it as is